transport-native-epoll/src/test/java/io/netty/channel/libaio/test/DirectFileDescriptorTest.java

```diff
@@ -0,0 +1,380 @@
+            exceptionThrown = true;
+        }
+
+        Assert.assertTrue(exceptionThrown);
+
+        fileDescriptor = control.newFile(temporaryFolder.newFile());
+
+        ByteBuffer buffer = fileDescriptor.newBuffer(512);
+
+        for (int i = 0; i < 512; i++) {
+            buffer.put((byte) 'a');
+        }
+
+        exceptionThrown = false;
+        try {
+            for (int i = 0; i < 60; i++) {
+                fileDescriptor.write(i * 512, 512, buffer, new CountClass());
+            }
+        } catch (IOException e) {
+            exceptionThrown = true;
+        }
+
+        Assert.assertTrue(exceptionThrown);
+
+        Object[] callbacks = new Object[100];
+        Assert.assertEquals(50, control.poll(callbacks, 50, 100));
+
+        // it should be possible to write now after queue space being released
+        fileDescriptor.write(0, 512, buffer, new CountClass());
+        Assert.assertEquals(1, control.poll(callbacks, 1, 100));
+
+        CountClass errorCallback = new CountClass();
+        // odd positions will have failures through O_DIRECT
+        fileDescriptor.read(3, 512, buffer, errorCallback);
+        Assert.assertEquals(1, control.poll(callbacks, 1, 50));
+        Assert.assertTrue(callbacks[0] instanceof ErrorInfo);
+        Assert.assertSame(errorCallback, ((ErrorInfo) callbacks[0]).callback());
+
+        // to help GC and the checkLeaks
+        callbacks = null;
+        errorCallback = null;
+
+        CountClass.checkLeaks();
+
+        exceptionThrown = false;
+        try {
+            DirectFileDescriptorController.newAlignedBuffer(300, 512);
+        } catch (RuntimeException e) {
+            exceptionThrown = true;
+        }
+
+        Assert.assertTrue(exceptionThrown);
+
+        exceptionThrown = false;
+        try {
+            DirectFileDescriptorController.newAlignedBuffer(-512, 512);
+        } catch (RuntimeException e) {
+            exceptionThrown = true;
+        }
+
+        Assert.assertTrue(exceptionThrown);
+
+        DirectFileDescriptorController.freeBuffer(buffer);
+    }
+
+    private void fillupFile(File file, int blocks) throws IOException {
+        FileOutputStream fileOutputStream = new FileOutputStream(file);
+        byte[] bufferWrite = new byte[512];
+        for (int i = 0; i < 512; i++) {
+            bufferWrite[i] = (byte) 'A';
+        }
+
+        for (int i = 0; i < blocks; i++) {
+            fileOutputStream.write(bufferWrite);
+        }
+
+        fileOutputStream.close();
+    }
+
+    static class CountClass {
+        static AtomicInteger count = new AtomicInteger(0);
+
+        public CountClass() {
+            count.incrementAndGet();
+        }
+
+        @Override
+        protected void finalize() throws Throwable {
+            super.finalize();
+            count.decrementAndGet();
+        }
+
+        public static void checkLeaks() throws InterruptedException {
+            long timeout = System.currentTimeMillis() + 5000;
```

Comment:
I just want a timeout for 5 seconds. if the objects are not released for 5 seconds then it would throw a failure through the assertEquals. 

I think this is sufficient.